torch/csrc/dynamo/eval_frame.c

```diff
@@ -1018,6 +1018,17 @@ static PyObject* _custom_eval_frame(
     // reference seems wrong. Therefore, we directly access the
     // extra->cache_entry. extra wont be NULL here.
     extra->cache_entry = create_cache_entry(extra->cache_entry, result);
+
+    // The frame stores the locals in fast representation (frame->localsplus).
+    // To simplify the implementation of lookup etc, we use
+    // PyFrame_FastToLocalsWithError to create a dictionary from the fast
+    // representation. However, this dictionary can hold the references of
+    // keys/values which can lead to memory leak (see
+    // test_release_input_memory). Therefore, we clear the dictionary here. The
+    // later call to eval_custom_code uses the fast representation for accessing
+    // the locals.
+    PyDict_Clear(frame->f_locals);
```

Comment:
I don't think this is right.  My understanding is:
1) `frame->f_locals` starts as `NULL`
2) `PyFrame_FastToLocalsWithError` changes it to a `dict`
3) Here we should change it back to `NULL`

Something like:
```
Py_XDECREF(frame->f_locals);
frame->f_locals = NULL;
```

Reply:
I checked and yes `frame->f_locals` starts as `NULL`. So, your solution makes sense.

But this test still fails
https://github.com/pytorch/pytorch/blob/2ed0af2bdeb2a5e0df78b22eab04d468f8db239e/test/dynamo/test_misc.py#L4067-L4080

Which likely means someone else is also holding on to f_locals or its contents in some way.

Reply:
Is there a pointer sticking around in the compiler code somewhere?

Perhaps use, `gc.get_referrers()` to find them,
https://docs.python.org/3/library/gc.html#gc.get_referrers